File: server-pedidos-java/src/main/java/br/edu/utfpr/serverpedidosjava/controller/CategoriaController.java
package br.edu.utfpr.serverpedidosjava.controller;

import br.edu.utfpr.serverpedidosjava.controller.dto.CategoriaDTO;
import br.edu.utfpr.serverpedidosjava.entity.Categoria;
import br.edu.utfpr.serverpedidosjava.repository.CategoriaRepository;
import jakarta.annotation.Nullable;
import jakarta.validation.Valid;
import org.apache.logging.log4j.util.Strings;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/categorias")
public class CategoriaController {

    private final CategoriaRepository categoriaRepository;

    public CategoriaController(CategoriaRepository categoriaRepository) {
        this.categoriaRepository = categoriaRepository;
    }

    @GetMapping
    public List<CategoriaDTO> list(@RequestParam(name = "nome") @Nullable String nome) {
        List<Categoria> categorias;
        if (Strings.isNotBlank(nome)) {
            categorias = this.categoriaRepository.findByNomeContainingIgnoreCase(nome);
        } else {
            categorias = this.categoriaRepository.findAll();
        }
        return categorias.stream().map(CategoriaDTO::fromEntity).toList();
    }

    @GetMapping("/page")
    public Page<CategoriaDTO> paginar(@RequestParam(name = "nome") @Nullable String nome, Pageable pageable) {
        Page<Categoria> categorias;
        if (Strings.isNotBlank(nome)) {
            categorias = this.categoriaRepository.findByNomeContainingIgnoreCase(nome, pageable);
        } else {
            categorias = this.categoriaRepository.findAll(pageable);
        }
        return categorias.map(CategoriaDTO::fromEntity);
    }

    @GetMapping("/{id}")
    public ResponseEntity<CategoriaDTO> findById(@PathVariable(name = "id") Integer id) {
        Categoria categoria = this.categoriaRepository.findById(id).orElse(null);
        if (categoria == null) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.ok(CategoriaDTO.fromEntity(categoria));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable(name = "id") Integer id) {
        this.categoriaRepository.findById(id).ifPresent(categoriaRepository::delete);
        return ResponseEntity.noContent().build();
    }

    @PostMapping
    public CategoriaDTO save(@Valid @RequestBody CategoriaDTO categoria) {
        Categoria categoriaAtualizada = this.categoriaRepository.save(categoria.toEntity());
        return CategoriaDTO.fromEntity(categoriaAtualizada);
    }
}

File: server-pedidos/src/main/kotlin/br/edu/utfpr/serverpedidos/controller/CategoriaController.kt
package br.edu.utfpr.serverpedidos.controller

import br.edu.utfpr.serverpedidos.controller.dto.CategoriaDTO
import br.edu.utfpr.serverpedidos.repository.CategoriaRepository
import jakarta.validation.Valid
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import kotlin.jvm.optionals.getOrNull

@RestController
@RequestMapping("/categorias")
class CategoriaController(
    private val categoriaRepository: CategoriaRepository
) {

    @GetMapping
    fun list(@RequestParam nome: String?): List<CategoriaDTO> {
        val categorias = nome?.let {
            categoriaRepository.findByNomeContainsIgnoreCase(it)
        } ?: categoriaRepository.findAll()
        return categorias.map { CategoriaDTO.fromEntity(it) }
    }

    @GetMapping("/page")
    fun paginar(@RequestParam nome: String?, pageable: Pageable): Page<CategoriaDTO> {
        val categorias = nome?.let {
            categoriaRepository.findByNomeContainsIgnoreCase(it, pageable)
        } ?: categoriaRepository.findAll(pageable)
        return categorias.map { CategoriaDTO.fromEntity(it) }
    }

    @GetMapping("/{id}")
    fun findById(@PathVariable id: Int): ResponseEntity<CategoriaDTO> {
        val categoria = categoriaRepository.findById(id).getOrNull()
            ?: return ResponseEntity
                .notFound()
                .build()
        return ResponseEntity.ok(CategoriaDTO.fromEntity(categoria))
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: Int): ResponseEntity<Unit> {
        categoriaRepository.findById(id).getOrNull()?.let {
            categoriaRepository.delete(it)
        }
        return ResponseEntity
            .noContent()
            .build()
    }

    @PostMapping
    fun save(@Valid @RequestBody categoria: CategoriaDTO): CategoriaDTO {
        val categoriaAtualizada = categoriaRepository.save(categoria.toEntity())
        return CategoriaDTO.fromEntity(categoriaAtualizada)
    }
}
